modules/table/src/main/java/org/apache/ignite/internal/table/distributed/disaster/PartitionStatesMetricSource.java

```diff
@@ -0,0 +1,152 @@
+    ) {
+        this.tableId = tableDescriptor.id();
+        this.disasterRecoveryManager = disasterRecoveryManager;
+
+        metricSourceName = String.format("partition.states.zone.%s.table.%s", tableDescriptor.zoneId(), tableDescriptor.id());
+    }
+
+    @Override
+    public String name() {
+        return metricSourceName;
+    }
+
+    @Override
+    public synchronized @Nullable MetricSet enable() {
+        if (enabled) {
+            return null;
+        }
+
+        var builder = new MetricSetBuilder(metricSourceName);
+
+        builder.longGauge(
+                "UnavailablePartitionCount",
+                "Count of partitions not yet started.",
+                () -> calculatePartitionCountByLocalState(UNAVAILABLE)
+        );
+
+        builder.longGauge(
+                "HealthyPartitionCount",
+                "Count of living partitions with a healthy state machine.",
+                () -> calculatePartitionCountByLocalState(HEALTHY)
+        );
+
+        builder.longGauge(
+                "InitializingPartitionCount",
+                "Count of partitions that are starting right now.",
+                () -> calculatePartitionCountByLocalState(INITIALIZING)
+        );
+
+        builder.longGauge(
+                "InstallingSnapshotPartitionCount",
+                "Count of partitions that installing Raft snapshots from the leader.",
+                () -> calculatePartitionCountByLocalState(INSTALLING_SNAPSHOT)
+        );
+
+        builder.longGauge(
+                "BrokenPartitionCount",
+                "Count of broken partitions.",
+                () -> calculatePartitionCountByLocalState(BROKEN)
+        );
+
+        return builder.build();
+    }
+
+    @Override
+    public synchronized void disable() {
+        enabled = false;
+    }
+
+    @Override
+    public synchronized boolean enabled() {
+        return enabled;
+    }
+
+    private long calculatePartitionCountByLocalState(LocalPartitionStateEnum state) {
+        long[] count = {0};
+
+        disasterRecoveryManager.raftManager.forEach((raftNodeId, raftGroupService) -> {
+            if (raftNodeId.groupId() instanceof TablePartitionId) {
+                var tablePartitionId = (TablePartitionId) raftNodeId.groupId();
+
+                if (tablePartitionId.tableId() == tableId) {
+                    LocalPartitionStateEnum partitionState = localPartitionState(raftGroupService);
+
+                    if (partitionState == state) {
+                        count[0]++;
+                    }
+                }
+            }
+        });
+
+        return count[0];
+    }
+
+    private static LocalPartitionStateEnum localPartitionState(RaftGroupService raftGroupService) {
+        Node raftNode = raftGroupService.getRaftNode();
+
+        LocalPartitionStateEnum localState = convertState(raftNode.getNodeState());
+        long lastLogIndex = raftNode.lastLogIndex();
+
+        if (localState == HEALTHY) {
```

Comment:
Looks like a copy-paste, maybe you should re-use existing code instead. If not, then please explain you reasoning, thank you

Reply:
I agree that it looks like a copy-paste, I’ll try to look at it again or comment.

Reply:
try fix it